src/System.Private.Xml/src/System/Xml/Base64Decoder.cs

```diff
@@ -0,0 +1,240 @@
+            fixed (char* pChars = &chars[startPos])
+            {
+                fixed (byte* pBytes = &_buffer[_curIndex])
+                {
+                    Decode(pChars, pChars + len, pBytes, pBytes + (_endIndex - _curIndex), out charsDecoded, out bytesDecoded);
+                }
+            }
+            _curIndex += bytesDecoded;
+            return charsDecoded;
+        }
+
+        [System.Security.SecuritySafeCritical]
+        internal override unsafe int Decode(string str, int startPos, int len)
+        {
+            if (str == null)
+            {
+                throw new ArgumentNullException(nameof(str));
+            }
+            if (len < 0)
+            {
+                throw new ArgumentOutOfRangeException(nameof(len));
+            }
+            if (startPos < 0)
+            {
+                throw new ArgumentOutOfRangeException(nameof(startPos));
+            }
+            if (str.Length - startPos < len)
+            {
+                throw new ArgumentOutOfRangeException(nameof(len));
+            }
+
+            if (len == 0)
+            {
+                return 0;
+            }
+            int bytesDecoded, charsDecoded;
+            fixed (char* pChars = str)
+            {
+                fixed (byte* pBytes = &_buffer[_curIndex])
+                {
+                    Decode(pChars + startPos, pChars + startPos + len, pBytes, pBytes + (_endIndex - _curIndex), out charsDecoded, out bytesDecoded);
+                }
+            }
+            _curIndex += bytesDecoded;
+            return charsDecoded;
+        }
+
+        internal override void Reset()
+        {
+            _bitsFilled = 0;
+            _bits = 0;
+        }
+
+        internal override void SetNextOutputBuffer(Array buffer, int index, int count)
+        {
+            Debug.Assert(buffer != null);
+            Debug.Assert(count >= 0);
+            Debug.Assert(index >= 0);
+            Debug.Assert(buffer.Length - index >= count);
+            Debug.Assert((buffer as byte[]) != null);
+
+            _buffer = (byte[])buffer;
+            _startIndex = index;
+            _curIndex = index;
+            _endIndex = index + count;
+        }
+
+        //
+        // Private methods
+        //
+        private static byte[] ConstructMapBase64()
+        {
+            byte[] mapBase64 = new byte[MaxValidChar + 1];
+            for (int i = 0; i < mapBase64.Length; i++)
+            {
+                mapBase64[i] = Invalid;
+            }
+            for (int i = 0; i < s_charsBase64.Length; i++)
+            {
+                mapBase64[(int)s_charsBase64[i]] = (byte)i;
+            }
+            return mapBase64;
+        }
+
+        [System.Security.SecurityCritical]
+        private unsafe void Decode(char* pChars, char* pCharsEndPos,
+                             byte* pBytes, byte* pBytesEndPos,
+                             out int charsDecoded, out int bytesDecoded)
+        {
+#if DEBUG
```

Comment:
This is a port of desktop code, I'd leave those kind of improvement as a post-step